critwall/src/main/java/org/opencadc/critwall/FileSyncJob.java

```diff
@@ -241,41 +244,59 @@ public void run() {
      * @throws InterruptedException
      * @throws WriteException
      */
-    private StorageMetadata syncArtifact(Iterator<URL> urlIterator, URI artifact, StorageAdapter sa)
-        throws StorageEngageException, InterruptedException, WriteException {
+    private StorageMetadata syncArtifact(Iterator<URL> urlIterator, URI artifact, StorageAdapter sa, ArtifactDAO artDao)
+        throws StorageEngageException, InterruptedException, WriteException, IllegalArgumentException {
 
         StorageMetadata storageMeta = null;
 
-        while (urlIterator.hasNext()) {
-            URL u = urlIterator.next();
-            log.debug("trying " + u);
-
-            try {
-                ByteArrayOutputStream dest = new ByteArrayOutputStream();
-                HttpGet get = new HttpGet(u, dest);
-                get.prepare();
-
-                NewArtifact a = new NewArtifact(artifact);
-                storageMeta = sa.put(a, get.getInputStream());
-                log.debug("storage meta returned: " + storageMeta.getStorageLocation());
-                return storageMeta;
-
-            } catch (WriteException wre) {
-                // IOException will capture this if not explicitly caught
-                log.info("write error for storage adapter put: " + wre.getMessage());
-                throw wre;
-            } catch (TransientException | IOException te) {
-                // ReadException will be caught under the IOException
-                // could be prepare or put throwing this error
-                // will move to next url
-                log.info("transient exception." + te.getMessage());
-
-            } catch (ResourceNotFoundException | ResourceAlreadyExistsException  badURL) {
-                log.info("removing URL " + u + " from list: " + badURL.getMessage());
-                urlIterator.remove();
-
+        if (urlIterator.hasNext()) {
+            while (urlIterator.hasNext()) {
+                URL u = urlIterator.next();
+                log.debug("trying " + u);
+
+                try {
+                    ByteArrayOutputStream dest = new ByteArrayOutputStream();
+                    HttpGet get = new HttpGet(u, dest);
+                    get.prepare();
+
+                    // Check content checksum and length to verify this
+                    // artifact can be synced currently
+                    Artifact curArtifact = artDao.get(artifact);
+                    log.debug("content checksums: " + get.getContentMD5() + "\n" + curArtifact.getContentChecksum().getSchemeSpecificPart());
```

Comment:
These checks here will work out if the remote server provides the values in the header metadata, but they are optional so get.getContentMD5() and get.getContentLength() can return nulls. The same methods in curArtifact never return null (see data model).

So these checks need to be tweaked to avoid NullPointerException.

Whether or not the `get` provides values for contentMD5 or contentLength, the values from curArtifact ***must*** be included with `NewArtifact a` below so that the storage adapter can verify the downloaded data actually matches the expectation. That protects against corrupted downloads and as well as the case where one of the sites has a different copy of the file (either newer or older - no way to tell here).

Technically, you could not do any checks here and rely on the storage adapter and it would get the correct outcome. However, the adapter would have to download the whole file in order to fail so the checks here are a optimisation. Please add a comment to that effect.